### PR TITLE
drivers: usb: Add nRF52840 USBD driver

### LINKED ISSUE #9628: nRF52x USB device stack support
I want to integrate nRF52840-PCA10056 board with WebUSB. 
Does nRF52x USB device stack is supported by Zephyr?

### PR DESCRIPTION
**Pull request by pawelzadrozniak**
* Add usbd_dc_nrfx shim
* Add nrfx_usbd driver into ext/hal/nordic/nrfx

The shim is based on the previous one usbd_dc_nrf5.
For handling the USBD hardware, tested nrfx_usbd driver from nRF SDK
was used.

Briefly tested examples:
* usb/cdc_acm
* usb/dfu (USB communication only due to flash handling issues)
* usb/hid-mouse
* bluetooth/hci_usb

The driver relies on nrfx (ext/hal/nordic/nrfx) which should be updated/merged in the first place.

Fixes #9628.

### CODE REVIEW

File: drivers/usb/device/usb_dc_nrfx.c [hunk 1]
```diff
@@ -0,0 +1,1580 @@
...
+
+			SYS_LOG_DBG("read request, ep %d",
+				    (u32_t)p_event->data.eptransfer.ep);
+
+			ev->ep = ep_ctx;
+			ev->evt = EP_EVT_RECV_REQ;
+			usbd_evt_put(ev);
+			ep_ctx->read_pending = true;
+			usbd_work_schedule();
+		} else if (p_event->data.eptransfer.status == NRF_USBD_EP_OK) {
```
Comment (anangl) [inline drivers/usb/device/usb_dc_nrfx.c]: And what about the other statuses: `NRF_USBD_EP_OVERLOAD` and `NRF_USBD_EP_ABORTED`. Shouldn't these also be handled (at least regarding `ctx->dma_in_use`)?

**Comment by jfischer-phytec-iot** [comment]
@pawelzadrozniak please rebase on master (rebased on my own I got two issues: NRF_USBD_ISOSPLIT_Half is NRF_USBD_ISOSPLIT_HALF now and [usb/dc] [ERR] hf_clock_enable: NRF5 HF clock start fail: -68)

**Comment by carlescufi** [comment]
Depends on #10148, rebase after that is merged

**Comment by carlescufi** [comment]
hey @pawelzadrozniak, #10148 is merged. Can you rebase this?

**Comment by carlescufi** [comment]
@pawelzadrozniak there are some checkpatch errors here:
https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/23037/5/tests

@finikorg @jfischer-phytec-iot could you take a look at this please?

**Comment by carlescufi** [comment]
@rakons would you mind revisiting this PR now that it has been rebased by @pawelzadrozniak ?

**Review by rakons — CHANGES_REQUESTED** [review]

File: drivers/usb/device/usb_dc_nrfx.c [hunk 2]
```diff
@@ -0,0 +1,1630 @@
...
+
+	usbd_evt_flush();
+	k_sem_reset(&ctx->dma_in_use);
+	endpoint_ctx_deinit();
+
+	/* Following functions may return error if the peripheral
+	 * was not initialized before. This can be ignored.
+	 */
+	(void)nrfx_usbd_disable();
+	(void)nrfx_usbd_uninit();
```
Comment (anangl) [inline drivers/usb/device/usb_dc_nrfx.c]: `nrfx_usbd_uninit()` no longer returns a value.

File: ext/hal/nordic/nrfx_config_nrf52840.h [hunk 3]
```diff
@@ -2977,7 +2977,7 @@
 // <e> NRFX_USBD_ENABLED - nrfx_usbd - USBD peripheral driver
 //==========================================================
 #ifndef NRFX_USBD_ENABLED
-#define NRFX_USBD_ENABLED 0
+#define NRFX_USBD_ENABLED 1
 #endif
```
Comment (anangl) [inline ext/hal/nordic/nrfx_config_nrf52840.h]:
```
#ifdef CONFIG_NRFX_USBD
#define NRFX_USBD_ENABLED 1
#endif
```

File: ext/hal/nordic/nrfx_config_nrf52840.h [hunk 4]
```diff
@@ -2163,7 +2163,7 @@
 
 
 #ifndef NRFX_SYSTICK_ENABLED
-#define NRFX_SYSTICK_ENABLED 0
+#define NRFX_SYSTICK_ENABLED 1
 #endif
```
[the remark this answers is not in the document]
Comment (anangl) [reply ext/hal/nordic/nrfx_config_nrf52840.h]: Still not addressed.

**Review by anangl — DISMISSED** [review]

**Comment by carlescufi** [comment]
@pawelzadrozniak the commit message in the first commit is wrong, since you are no longer adding anything to `ext/`

File: drivers/usb/device/usb_dc_nrfx.c [hunk 5]
```diff
@@ -0,0 +1,1567 @@
...
+#include <usb/usb_device.h>
+#include <clock_control.h>
+#include <nrf_power.h>
+#include <drivers/clock_control/nrf5_clock_control.h>
+
+#include <drivers/include/nrfx_usbd.h>
+
+
+
+
```
Comment (finikorg) [inline drivers/usb/device/usb_dc_nrfx.c]: Those newlines are not needed here and below

File: drivers/usb/device/usb_dc_nrfx.c [hunk 6]
```diff
@@ -0,0 +1,1567 @@
...
+			return &ctx->ep_ctx[CFG_EPIN_CNT +
+					    CFG_EP_ISOIN_CNT +
+					    ep_num];
+		}
+	}
+
+	return NULL;
+}
+
+
```
Comment (finikorg) [inline drivers/usb/device/usb_dc_nrfx.c]: one newline is enough

File: subsys/usb/usb_device.c [hunk 7]
```diff
@@ -234,7 +234,7 @@ static bool usb_handle_request(struct usb_setup_packet *setup,
  */
 static void usb_data_to_host(void)
 {
-	u32_t chunk = min(MAX_PACKET_SIZE0, usb_dev.data_buf_residue);
+	u32_t chunk = usb_dev.data_buf_residue;
```
Comment (anangl) [inline subsys/usb/usb_device.c]: This change should be done in a separate commit (currently it is in "drivers: usb: Add nRF52840 USBD driver") and described properly in the commit message.

File: subsys/usb/usb_device.c [hunk 8]
```diff
@@ -234,7 +234,7 @@ static bool usb_handle_request(struct usb_setup_packet *setup,
...
 
 	/*Always EP0 for control*/
 	usb_dc_ep_write(0x80, usb_dev.data_buf, chunk, &chunk);
```
[the remark this answers is not in the document]
Comment (jfischer-phytec-iot) [reply subsys/usb/usb_device.c]: Is not part of this PR and there is nothing magic in 0x80

File: subsys/usb/usb_device.c (hunk 7 above)
Comment (jfischer-phytec-iot) [reply subsys/usb/usb_device.c]: Maybe it would be better to open own PR for it, I do not sure that every driver can handle it. We should test existing drivers before it is merged.

**Review by jfischer-no — COMMENTED** [review]
I wanted to test it with sample/subsys/usb/testusb, but it fails on set config. I will check with the sniffer later.

```
[194772.826699] usbcore: registered new interface driver usbtest
[194779.379240] usb 9-7.4: new full-speed USB device number 49 using ehci-pci
[194779.490653] usb 9-7.4: New USB device found, idVendor=2fe3, idProduct=0100
[194779.490657] usb 9-7.4: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[194779.490660] usb 9-7.4: Product: Zephyr testusb sample
[194779.490663] usb 9-7.4: Manufacturer: ZEPHYR
[194779.490666] usb 9-7.4: SerialNumber: 0.01
[194784.602544] usb 9-7.4: can't set config #1, error -110
```

File: subsys/usb/usb_device.c (hunk 7 above)
Comment (carlescufi) [reply subsys/usb/usb_device.c]: @jfischer-phytec-iot  @finikorg I think that's probably the safest.
@pawelzadrozniak can you please open a separate PR with this change so we can get that tested and merged separately?

**Comment by carlescufi** [comment]
@rakons @anangl can you re-review without taking into account the extra commit that fixes a USB core issue?

**Review by carlescufi — APPROVED** [review]

**Comment by nashif** [comment]
@pawelzadrozniak please rebase

File: drivers/clock_control/nrf5_power_clock.c [hunk 9]
```diff
@@ -414,70 +414,32 @@ DEVICE_AND_API_INIT(clock_nrf5_k32src,
...
 
-	if (enable) {
-		nrf_power_int_enable(mask);
-	} else {
-		nrf_power_int_disable(mask);
-	}
-}
-
-static bool usbregstatus_vbusdet_get(void)
+bool nrf5_power_clock_usb_vbusdet(void)
```
Comment (anangl) [inline drivers/clock_control/nrf5_power_clock.c]: I see no point in keeping this function. The USB shim can call the POWER HAL itself. Same applies to `nrf5_power_clock_usb_outrdy`.
`nrf5_power_usb_power_int_enable` is a bit different since it refers to the POWER_CLOCK IRQ, so it should stay here (temporarily, until we find a better solution instead of it),

**Comment by carlescufi** [comment]
@anangl could you take one more look at this?

**Review by carlescufi — APPROVED** [review]
@pawelzadrozniak you are changing files in `ext/hal` in the same commit where you add the driver shim. Can you split the changes to `ext/hal` in a separate commit?
Also can you please merge the last commit, the one increasing FIFO_ELEM_COUNT into the driver shim commit?

**Review by carlescufi — CHANGES_REQUESTED** [review]
Sorry, meant to request changes

File: drivers/usb/device/Kconfig [hunk 10]
```diff
@@ -53,6 +53,8 @@ config USB_NRF52840
 	depends on SOC_NRF52840
 	select USB_DEVICE_DRIVER
 	select HAS_DTS_USB
+	select NRFX_USBD
+	select NRFX_SYSTICK
```
Comment (anangl) [inline drivers/usb/device/Kconfig]: This should be selected by the NRFX_USBD option itself. Just like NRFX_NFCT selects NRFX_TIMER4.

File: drivers/usb/device/usb_dc_nrfx.c [hunk 11]
```diff
@@ -0,0 +1,1560 @@
...
+	k_sem_take(&ctx->dma_in_use, K_FOREVER);
+	NRFX_USBD_TRANSFER_OUT(transfer, ep_ctx->buf.data,
+			       ep_ctx->cfg.max_sz);
+	nrfx_err_t err = nrfx_usbd_ep_transfer(
+		ep_addr_to_nrfx(ep_ctx->cfg.addr), &transfer);
+	if (err != NRFX_SUCCESS) {
+		SYS_LOG_ERR(
+			"nRF USBD transfer error (OUT): %d.",
+			err);
```
Comment (anangl) [inline drivers/usb/device/usb_dc_nrfx.c]: You could improve the formatting. Now there is no need to break this line into three pieces. Similar story in [line 572](https://github.com/zephyrproject-rtos/zephyr/pull/8787/files#diff-b6600658d4d8775791cddce8653f308cR572).

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@anangl one last pass?

**Review by anangl — APPROVED** [review]
